### PR TITLE
Heuristic compression mode selector

### PR DESCRIPTION
**Pull request by m1awicki**
### Description
This change introduces heuristic compression mode selection mechanism. It is based on patch and publication of Jae-Ho Nah.

### General diff between images compressed using new and old compression mode selection mechanism:
result23
result03
result05
result20

### Detailed comparison of images compressed using new compression mode selector (_new_) with current implementation (_old_):
kodim03_cmp Image: kodim03.png

kodim20_cmp Image: kodim20.png

### CODE REVIEW

File: ProcessRGB.cpp [hunk 1]
```diff
@@ -1979,13 +2008,153 @@ static etcpak_force_inline uint64_t ProcessRGB( const uint8_t* src )
...
+    static const __m128i interleaving_mask_hi = _mm_set_epi8( 14, 12, 10, 8, 6, 4, 2, 0, 15, 13, 11, 9, 7, 5, 3, 1 );
+    __m128i luma_8bit_lo_moved = _mm_shuffle_epi8( luma_8bit_lo, interleaving_mask_lo );
+    __m128i luma_8bit_hi_moved = _mm_shuffle_epi8( luma_8bit_hi, interleaving_mask_hi );
+    __m128i luma_8bit = _mm_or_si128( luma_8bit_hi_moved, luma_8bit_lo_moved );
+    luma.luma8 = luma_8bit;
+
+    // min/max calculation
+    luma.min = hMin( luma_8bit, luma.minIdx ) * 0.00392156f;
+    luma.max = hMax( luma_8bit, luma.maxIdx ) * 0.00392156f;
+    float luma_range = luma.max - luma.min;
```
Comment (wolfpld) [inline ProcessRGB.cpp:2079]: This is unused.

Comment (m1awicki) [reply ProcessRGB.cpp]: Ah, right. Removed it.

File: ProcessRGB.cpp [hunk 2]
```diff
@@ -1979,13 +2008,155 @@ static etcpak_force_inline uint64_t ProcessRGB( const uint8_t* src )
...
+    }
+#endif
+}
+
+static etcpak_force_inline uint8_t SelectModeETC2( const Luma& luma )
+{
+#ifdef __AVX2__
+    const float lumaRange = ( luma.max - luma.min );
+#else
+    const float lumaRange = ( luma.max - luma.min ) / 255.f;
```
Comment (wolfpld) [inline ProcessRGB.cpp:2102]: Change this to multiply.

Comment (wolfpld) [reply ProcessRGB.cpp:2102]: https://gcc.godbolt.org/z/K9W17sGcM

**Comment by wolfpld** [comment]
Some performance numbers (Mpx/s):

| | ST | MT | HST | HMT |
|---|---|---|---|---|
|ETC1|407|6790|408|6699|
|ETC2|272|4640|397|6532|
|RGBA|207|3395|285|4351|

ST - single thread
MT - multi thread
H - heuristic

**Pull request merged**